In [1]:
import numpy as np
import pandas as pd
pd.set_option('display.max_columns', None)

In [2]:
df_eda = pd.read_csv('data.csv')

In [3]:
df_eda.head()

Description         Name  year iso_code  population  gdp  cement_co2  \
0     Country  Afghanistan  1850      AFG   3752993.0  NaN         0.0   
1     Country  Afghanistan  1851      AFG   3767956.0  NaN         0.0   
2     Country  Afghanistan  1852      AFG   3783940.0  NaN         0.0   
3     Country  Afghanistan  1853      AFG   3800954.0  NaN         0.0   
4     Country  Afghanistan  1854      AFG   3818038.0  NaN         0.0   

   cement_co2_per_capita  co2  co2_growth_abs  co2_growth_prct  \
0                    0.0  NaN             NaN              NaN   
1                    0.0  NaN             NaN              NaN   
2                    0.0  NaN             NaN              NaN   
3                    0.0  NaN             NaN              NaN   
4                    0.0  NaN             NaN              NaN   

   co2_including_luc  co2_including_luc_growth_abs  \
0                NaN                           NaN   
1                NaN                           NaN   
2                NaN                           NaN   
3                NaN                           NaN   
4                NaN                           NaN   

   co2_including_luc_growth_prct  co2_including_luc_per_capita  \
0                            NaN                           NaN   
1                            NaN                           NaN   
2                            NaN                           NaN   
3                            NaN                           NaN   
4                            NaN                           NaN   

   co2_including_luc_per_gdp  co2_including_luc_per_unit_energy  \
0                        NaN                                NaN   
1                        NaN                                NaN   
2                        NaN                                NaN   
3                        NaN                                NaN   
4                        NaN                                NaN   

   co2_per_capita  co2_per_gdp  co2_per_unit_energy  coal_co2  \
0             NaN          NaN                  NaN       NaN   
1             NaN          NaN                  NaN       NaN   
2             NaN          NaN                  NaN       NaN   
3             NaN          NaN                  NaN       NaN   
4             NaN          NaN                  NaN       NaN   

   coal_co2_per_capita  consumption_co2  consumption_co2_per_capita  \
0                  NaN              NaN                         NaN   
1                  NaN              NaN                         NaN   
2                  NaN              NaN                         NaN   
3                  NaN              NaN                         NaN   
4                  NaN              NaN                         NaN   

   consumption_co2_per_gdp  cumulative_cement_co2  cumulative_co2  \
0                      NaN                    0.0             NaN   
1                      NaN                    0.0             NaN   
2                      NaN                    0.0             NaN   
3                      NaN                    0.0             NaN   
4                      NaN                    0.0             NaN   

   cumulative_co2_including_luc  cumulative_coal_co2  cumulative_flaring_co2  \
0                           NaN                  NaN                     NaN   
1                           NaN                  NaN                     NaN   
2                           NaN                  NaN                     NaN   
3                           NaN                  NaN                     NaN   
4                           NaN                  NaN                     NaN   

   cumulative_gas_co2  cumulative_luc_co2  cumulative_oil_co2  \
0                 NaN               2.993                 NaN   
1                 NaN               6.008                 NaN   
2                 NaN               9.043                 NaN   
3                 NaN              12.095                 NaN

In [4]:
df_eda['Name'].nunique()

253

In [5]:
df_eda['Name'].unique()

array(['Afghanistan', 'Albania', 'Algeria', 'Andorra', 'Angola',
       'Antigua and Barbuda', 'Argentina', 'Armenia', 'Australia',
       'Austria', 'Azerbaijan', 'Bahamas', 'Bahrain', 'Bangladesh',
       'Barbados', 'Belarus', 'Belgium', 'Belize', 'Benin', 'Bhutan',
       'Bolivia', 'Bosnia and Herzegovina', 'Botswana', 'Brazil',
       'Brunei', 'Bulgaria', 'Burkina Faso', 'Burundi', 'Cambodia',
       'Cameroon', 'Canada', 'Cape Verde', 'Central African Republic',
       'Chad', 'Chile', 'China', 'Colombia', 'Comoros', 'Congo',
       'Cook Islands', 'Costa Rica', "Cote d'Ivoire", 'Croatia', 'Cuba',
       'Cyprus', 'Czechia', 'Democratic Republic of Congo', 'Denmark',
       'Djibouti', 'Dominica', 'Dominican Republic', 'East Timor',
       'Ecuador', 'Egypt', 'El Salvador', 'Equatorial Guinea', 'Eritrea',
       'Estonia', 'Eswatini', 'Ethiopia', 'Fiji', 'Finland', 'France',
       'Gabon', 'Gambia', 'Georgia', 'Germany', 'Ghana', 'Greece',
       'Grenada', 'Guatemala', 'Guine

In [6]:
df_eda['Name'].tail(150)

43596     International shipping
43597     International shipping
43598     International shipping
43599     International shipping
43600     International shipping
                  ...           
43741    International transport
43742    International transport
43743    International transport
43744    International transport
43745    International transport
Name: Name, Length: 150, dtype: object

In [7]:
df_eda.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 43746 entries, 0 to 43745
Data columns (total 80 columns):
 #   Column                                     Non-Null Count  Dtype  
---  ------                                     --------------  -----  
 0   Description                                43746 non-null  object 
 1   Name                                       43746 non-null  object 
 2   year                                       43746 non-null  int64  
 3   iso_code                                   37921 non-null  object 
 4   population                                 37965 non-null  float64
 5   gdp                                        14989 non-null  float64
 6   cement_co2                                 23612 non-null  float64
 7   cement_co2_per_capita                      22715 non-null  float64
 8   co2                                        27059 non-null  float64
 9   co2_growth_abs                             24987 non-null  float64
 10  co2_growth_prct       

In [8]:
df_eda['year']

0        1850
1        1851
2        1852
3        1853
4        1854
         ... 
43741    2019
43742    2020
43743    2021
43744    2022
43745    2023
Name: year, Length: 43746, dtype: int64

In [9]:
filter_year = df_eda[df_eda['year'] >= 2000] #hemos recortado por los 25 últimos años

In [10]:
filter_year.isna().sum()

Description                       0
Name                              0
year                              0
iso_code                        840
population                      552
                               ... 
temperature_change_from_n2o     792
total_ghg                       912
total_ghg_excluding_lucf        936
trade_co2                      2906
trade_co2_share                2906
Length: 80, dtype: int64

In [11]:
filter_year[filter_year['year'] > 2000].groupby('Name')['year'].count()

Name
Afghanistan          23
Africa               23
Africa (GCP)         23
Albania              23
Algeria              23
                     ..
Wallis and Futuna    23
World                23
Yemen                23
Zambia               23
Zimbabwe             23
Name: year, Length: 253, dtype: int64

In [12]:
nulos_porcentaje = filter_year.isna().sum() / filter_year.shape[0] * 100
nulos_porcentaje

Description                     0.000000
Name                            0.000000
year                            0.000000
iso_code                       13.833992
population                      9.090909
                                 ...    
temperature_change_from_n2o    13.043478
total_ghg                      15.019763
total_ghg_excluding_lucf       15.415020
trade_co2                      47.859025
trade_co2_share                47.859025
Length: 80, dtype: float64

In [13]:
filter_year['Name'].unique()


array(['Afghanistan', 'Albania', 'Algeria', 'Andorra', 'Angola',
       'Antigua and Barbuda', 'Argentina', 'Armenia', 'Australia',
       'Austria', 'Azerbaijan', 'Bahamas', 'Bahrain', 'Bangladesh',
       'Barbados', 'Belarus', 'Belgium', 'Belize', 'Benin', 'Bhutan',
       'Bolivia', 'Bosnia and Herzegovina', 'Botswana', 'Brazil',
       'Brunei', 'Bulgaria', 'Burkina Faso', 'Burundi', 'Cambodia',
       'Cameroon', 'Canada', 'Cape Verde', 'Central African Republic',
       'Chad', 'Chile', 'China', 'Colombia', 'Comoros', 'Congo',
       'Cook Islands', 'Costa Rica', "Cote d'Ivoire", 'Croatia', 'Cuba',
       'Cyprus', 'Czechia', 'Democratic Republic of Congo', 'Denmark',
       'Djibouti', 'Dominica', 'Dominican Republic', 'East Timor',
       'Ecuador', 'Egypt', 'El Salvador', 'Equatorial Guinea', 'Eritrea',
       'Estonia', 'Eswatini', 'Ethiopia', 'Fiji', 'Finland', 'France',
       'Gabon', 'Gambia', 'Georgia', 'Germany', 'Ghana', 'Greece',
       'Grenada', 'Guatemala', 'Guine

In [14]:
incluir = ['Afghanistan', 'Albania', 'Algeria', 'Andorra', 'Angola',
       'Antigua and Barbuda', 'Argentina', 'Armenia', 'Australia',
       'Austria', 'Azerbaijan', 'Bahamas', 'Bahrain', 'Bangladesh',
       'Barbados', 'Belarus', 'Belgium', 'Belize', 'Benin', 'Bhutan',
       'Bolivia', 'Bosnia and Herzegovina', 'Botswana', 'Brazil',
       'Brunei', 'Bulgaria', 'Burkina Faso', 'Burundi', 'Cambodia',
       'Cameroon', 'Canada', 'Cape Verde', 'Central African Republic',
       'Chad', 'Chile', 'China', 'Colombia', 'Comoros', 'Congo',
       'Cook Islands', 'Costa Rica', "Cote d'Ivoire", 'Croatia', 'Cuba',
       'Cyprus', 'Czechia', 'Democratic Republic of Congo', 'Denmark',
       'Djibouti', 'Dominica', 'Dominican Republic', 'East Timor',
       'Ecuador', 'Egypt', 'El Salvador', 'Equatorial Guinea', 'Eritrea',
       'Estonia', 'Eswatini', 'Ethiopia', 'Fiji', 'Finland', 'France',
       'Gabon', 'Gambia', 'Georgia', 'Germany', 'Ghana', 'Greece',
       'Grenada', 'Guatemala', 'Guinea', 'Guinea-Bissau', 'Guyana',
       'Haiti', 'Honduras', 'Hungary', 'Iceland', 'India', 'Indonesia',
       'Iran', 'Iraq', 'Ireland', 'Israel', 'Italy', 'Jamaica', 'Japan',
       'Jordan', 'Kazakhstan', 'Kenya', 'Kiribati', 'Kosovo', 'Kuwait',
       'Kyrgyzstan', 'Laos', 'Latvia', 'Lebanon', 'Lesotho', 'Liberia',
       'Libya', 'Liechtenstein', 'Lithuania', 'Luxembourg', 'Madagascar',
       'Malawi', 'Malaysia', 'Maldives', 'Mali', 'Malta',
       'Marshall Islands', 'Mauritania', 'Mauritius', 'Mexico',
       'Micronesia (country)', 'Moldova', 'Monaco', 'Mongolia',
       'Montenegro', 'Morocco', 'Mozambique', 'Myanmar', 'Namibia',
       'Nauru', 'Nepal', 'Netherlands', 'New Zealand', 'Nicaragua',
       'Niger', 'Nigeria', 'North Korea', 'North Macedonia', 'Norway',
       'Oman', 'Pakistan', 'Palau', 'Palestine', 'Panama',
       'Papua New Guinea', 'Paraguay', 'Peru', 'Philippines', 'Poland',
       'Portugal', 'Qatar', 'Romania', 'Russia', 'Rwanda',
       'Saint Kitts and Nevis', 'Saint Lucia',
       'Saint Vincent and the Grenadines', 'Samoa', 'San Marino',
       'Sao Tome and Principe', 'Saudi Arabia', 'Senegal', 'Serbia',
       'Seychelles', 'Sierra Leone', 'Singapore', 'Slovakia', 'Slovenia',
       'Solomon Islands', 'Somalia', 'South Africa', 'South Korea',
       'South Sudan', 'Spain', 'Sri Lanka', 'Sudan', 'Suriname', 'Sweden',
       'Switzerland', 'Syria', 'Taiwan', 'Tajikistan', 'Tanzania',
       'Thailand', 'Togo', 'Tonga', 'Trinidad and Tobago', 'Tunisia',
       'Turkey', 'Turkmenistan', 'Tuvalu', 'Uganda', 'Ukraine',
       'United Arab Emirates', 'United Kingdom', 'United States',
       'Uruguay', 'Uzbekistan', 'Vanuatu', 'Vatican', 'Venezuela',
       'Vietnam', 'Yemen', 'Zambia', 'Zimbabwe', 'Anguilla',
       'Antarctica', 'Aruba', 'Bermuda',
       'Bonaire Sint Eustatius and Saba', 'British Virgin Islands',
       'Christmas Island', 'Curacao', 'Faroe Islands', 'French Polynesia',
       'Greenland', 'Hong Kong', 'Macao', 'Montserrat', 'New Caledonia',
       'Niue', 'Saint Helena', 'Saint Pierre and Miquelon',
       'Sint Maarten (Dutch part)', 'Turks and Caicos Islands',
       'Wallis and Futuna' ]

In [15]:
df_filtrado = filter_year[filter_year["Name"].isin(incluir)]

In [16]:
df_filtrado.head(10)

Description         Name  year iso_code  population           gdp  \
150     Country  Afghanistan  2000      AFG  20130334.0  1.128379e+10   
151     Country  Afghanistan  2001      AFG  20284303.0  1.102127e+10   
152     Country  Afghanistan  2002      AFG  21378123.0  1.880487e+10   
153     Country  Afghanistan  2003      AFG  22733054.0  2.107434e+10   
154     Country  Afghanistan  2004      AFG  23560656.0  2.233257e+10   
155     Country  Afghanistan  2005      AFG  24404574.0  2.539769e+10   
156     Country  Afghanistan  2006      AFG  25424100.0  2.870440e+10   
157     Country  Afghanistan  2007      AFG  25909852.0  3.450753e+10   
158     Country  Afghanistan  2008      AFG  26482629.0  3.656104e+10   
159     Country  Afghanistan  2009      AFG  27466101.0  4.435872e+10   

     cement_co2  cement_co2_per_capita    co2  co2_growth_abs  \
150       0.010                  0.001  1.047          -0.045   
151       0.007                  0.000  1.069           0.022   
152       0.011                  0.001  1.341           0.272   
153       0.010                  0.000  1.560           0.219   
154       0.010                  0.000  1.237          -0.322   
155       0.006                  0.000  1.890           0.652   
156       0.012                  0.000  2.159           0.270   
157       0.012                  0.000  2.800           0.641   
158       0.015                  0.001  4.254           1.455   
159       0.013                  0.000  6.392           2.137   

     co2_growth_prct  co2_including_luc  co2_including_luc_growth_abs  \
150           -4.078              1.104                        -0.755   
151            2.098              1.840                         0.736   
152           25.432              6.115                         4.275   
153           16.302              5.925                        -0.190   
154          -20.669              3.352                        -2.572   
155           52.719              2.780                        -0.572   
156           14.279              3.128                         0.348   
157           29.666              3.565                         0.437   
158           51.951              4.887                         1.323   
159           50.239              6.904                         2.016   

     co2_including_luc_growth_prct  co2_including_luc_per_capita  \
150                        -40.599                         0.055   
151                         66.622                         0.091   
152                        232.400                         0.286   
153                         -3.113                         0.261   
154                        -43.420                         0.142   
155                        -17.055                         0.114   
156                         12.500                         0.123   
157                         13.962                         0.138   
158                         37.103                         0.185   
159                         41.256                         0.251   

     co2_including_luc_per_gdp  co2_including_luc_per_unit_energy  \
150                      0.098                              0.187   
151                      0.167                              0.394   
152                      0.325                              1.381   
153                      0.281                              1.138   
154                      0.150                              0.697   
155                      0.109                              0.452   
156                      0.109                              0.404   
157                      0.103                              0.388   
158                      0.134                              0.304   
159                      0.156                              0.259   

     co2_per_capita  co2_per_gdp  co2_per_unit_energy  coal_co2  \
150           0.052        0.093                0.177     0.004   
151           0.053        0.097     

In [17]:
diccionario_continentes = {
    'Afghanistan': 'Asia',
    'Albania': 'Europe',
    'Algeria': 'Africa',
    'Andorra': 'Europe',
    'Angola': 'Africa',
    'Antigua and Barbuda': 'North America',
    'Argentina': 'South America',
    'Armenia': 'Asia',
    'Australia': 'Oceania',
    'Austria': 'Europe',
    'Azerbaijan': 'Asia',
    'Bahamas': 'North America',
    'Bahrain': 'Asia',
    'Bangladesh': 'Asia',
    'Barbados': 'North America',
    'Belarus': 'Europe',
    'Belgium': 'Europe',
    'Belize': 'North America',
    'Benin': 'Africa',
    'Bhutan': 'Asia',
    'Bolivia': 'South America',
    'Bosnia and Herzegovina': 'Europe',
    'Botswana': 'Africa',
    'Brazil': 'South America',
    'Brunei': 'Asia',
    'Bulgaria': 'Europe',
    'Burkina Faso': 'Africa',
    'Burundi': 'Africa',
    'Cambodia': 'Asia',
    'Cameroon': 'Africa',
    'Canada': 'North America',
    'Cape Verde': 'Africa',
    'Central African Republic': 'Africa',
    'Chad': 'Africa',
    'Chile': 'South America',
    'China': 'Asia',
    'Colombia': 'South America',
    'Comoros': 'Africa',
    'Congo': 'Africa',
    'Cook Islands': 'Oceania',
    'Costa Rica': 'North America',
    "Cote d'Ivoire": 'Africa',
    'Croatia': 'Europe',
    'Cuba': 'North America',
    'Cyprus': 'Asia',
    'Czechia': 'Europe',
    'Democratic Republic of Congo': 'Africa',
    'Denmark': 'Europe',
    'Djibouti': 'Africa',
    'Dominica': 'North America',
    'Dominican Republic': 'North America',
    'East Timor': 'Asia',
    'Ecuador': 'South America',
    'Egypt': 'Africa',
    'El Salvador': 'North America',
    'Equatorial Guinea': 'Africa',
    'Eritrea': 'Africa',
    'Estonia': 'Europe',
    'Eswatini': 'Africa',
    'Ethiopia': 'Africa',
    'Fiji': 'Oceania',
    'Finland': 'Europe',
    'France': 'Europe',
    'Gabon': 'Africa',
    'Gambia': 'Africa',
    'Georgia': 'Asia',
    'Germany': 'Europe',
    'Ghana': 'Africa',
    'Greece': 'Europe',
    'Grenada': 'North America',
    'Guatemala': 'North America',
    'Guinea': 'Africa',
    'Guinea-Bissau': 'Africa',
    'Guyana': 'South America',
    'Haiti': 'North America',
    'Honduras': 'North America',
    'Hungary': 'Europe',
    'Iceland': 'Europe',
    'India': 'Asia',
    'Indonesia': 'Asia',
    'Iran': 'Asia',
    'Iraq': 'Asia',
    'Ireland': 'Europe',
    'Israel': 'Asia',
    'Italy': 'Europe',
    'Jamaica': 'North America',
    'Japan': 'Asia',
    'Jordan': 'Asia',
    'Kazakhstan': 'Asia',
    'Kenya': 'Africa',
    'Kiribati': 'Oceania',
    'Kosovo': 'Europe',
    'Kuwait': 'Asia',
    'Kyrgyzstan': 'Asia',
    'Laos': 'Asia',
    'Latvia': 'Europe',
    'Lebanon': 'Asia',
    'Lesotho': 'Africa',
    'Liberia': 'Africa',
    'Libya': 'Africa',
    'Liechtenstein': 'Europe',
    'Lithuania': 'Europe',
    'Luxembourg': 'Europe',
    'Madagascar': 'Africa',
    'Malawi': 'Africa',
    'Malaysia': 'Asia',
    'Maldives': 'Asia',
    'Mali': 'Africa',
    'Malta': 'Europe',
    'Marshall Islands': 'Oceania',
    'Mauritania': 'Africa',
    'Mauritius': 'Africa',
    'Mexico': 'North America',
    'Micronesia (country)': 'Oceania',
    'Moldova': 'Europe',
    'Monaco': 'Europe',
    'Mongolia': 'Asia',
    'Montenegro': 'Europe',
    'Morocco': 'Africa',
    'Mozambique': 'Africa',
    'Myanmar': 'Asia',
    'Namibia': 'Africa',
    'Nauru': 'Oceania',
    'Nepal': 'Asia',
    'Netherlands': 'Europe',
    'New Zealand': 'Oceania',
    'Nicaragua': 'North America',
    'Niger': 'Africa',
    'Nigeria': 'Africa',
    'North Korea': 'Asia',
    'North Macedonia': 'Europe',
    'Norway': 'Europe',
    'Oman': 'Asia',
    'Pakistan': 'Asia',
    'Palau': 'Oceania',
    'Palestine': 'Asia',
    'Panama': 'North America',
    'Papua New Guinea': 'Oceania',
    'Paraguay': 'South America',
    'Peru': 'South America',
    'Philippines': 'Asia',
    'Poland': 'Europe',
    'Portugal': 'Europe',
    'Qatar': 'Asia',
    'Romania': 'Europe',
    'Russia': 'Asia',
    'Rwanda': 'Africa',
    'Saint Kitts and Nevis': 'North America',
    'Saint Lucia': 'North America',
    'Saint Vincent and the Grenadines': 'North America',
    'Samoa': 'Oceania',
    'San Marino': 'Europe',
    'Sao Tome and Principe': 'Africa',
    'Saudi Arabia': 'Asia',
    'Senegal': 'Africa',
    'Serbia': 'Europe',
    'Seychelles': 'Africa',
    'Sierra Leone': 'Africa',
    'Singapore': 'Asia',
    'Slovakia': 'Europe',
    'Slovenia': 'Europe',
    'Solomon Islands': 'Oceania',
    'Somalia': 'Africa',
    'South Africa': 'Africa',
    'South Korea': 'Asia',
    'South Sudan': 'Africa',
    'Spain': 'Europe',
    'Sri Lanka': 'Asia',
    'Sudan': 'Africa',
    'Suriname': 'South America',
    'Sweden': 'Europe',
    'Switzerland': 'Europe',
    'Syria': 'Asia',
    'Taiwan': 'Asia',
    'Tajikistan': 'Asia',
    'Tanzania': 'Africa',
    'Thailand': 'Asia',
    'Togo': 'Africa',
    'Tonga': 'Oceania',
    'Trinidad and Tobago': 'North America',
    'Tunisia': 'Africa',
    'Turkey': 'Asia',
    'Turkmenistan': 'Asia',
    'Tuvalu': 'Oceania',
    'Uganda': 'Africa',
    'Ukraine': 'Europe',
    'United Arab Emirates': 'Asia',
    'United Kingdom': 'Europe',
    'United States': 'North America',
    'Uruguay': 'South America',
    'Uzbekistan': 'Asia',
    'Vanuatu': 'Oceania',
    'Vatican': 'Europe',
    'Venezuela': 'South America',
    'Vietnam': 'Asia',
    'Yemen': 'Asia',
    'Zambia': 'Africa',
    'Zimbabwe': 'Africa',
    'Anguilla': 'North America',
    'Antarctica': 'Antarctica',
    'Aruba': 'North America',
    'Bermuda': 'North America',
    'Bonaire Sint Eustatius and Saba': 'North America',
    'British Virgin Islands': 'North America',
    'Christmas Island': 'Asia',
    'Curacao': 'North America',
    'Faroe Islands': 'Europe',
    'French Polynesia': 'Oceania',
    'Greenland': 'North America',
    'Hong Kong': 'Asia',
    'Macao': 'Asia',
    'Montserrat': 'North America',
    'New Caledonia': 'Oceania',
    'Niue': 'Oceania',
    'Saint Helena': 'Africa',
    'Saint Pierre and Miquelon': 'North America',
    'Sint Maarten (Dutch part)': 'North America',
    'Turks and Caicos Islands': 'North America',
    'Wallis and Futuna': 'Oceania'
}

In [18]:
df_filtrado['Continent'] = df_filtrado['Name'].map(diccionario_continentes)


/var/folders/fw/63lcqvrj3gj5_78zkmp2ql8r0000gp/T/ipykernel_63810/839544734.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_filtrado['Continent'] = df_filtrado['Name'].map(diccionario_continentes)


In [19]:
df_filtrado.head(1)

Description         Name  year iso_code  population           gdp  \
150     Country  Afghanistan  2000      AFG  20130334.0  1.128379e+10   

     cement_co2  cement_co2_per_capita    co2  co2_growth_abs  \
150        0.01                  0.001  1.047          -0.045   

     co2_growth_prct  co2_including_luc  co2_including_luc_growth_abs  \
150           -4.078              1.104                        -0.755   

     co2_including_luc_growth_prct  co2_including_luc_per_capita  \
150                        -40.599                         0.055   

     co2_including_luc_per_gdp  co2_including_luc_per_unit_energy  \
150                      0.098                              0.187   

     co2_per_capita  co2_per_gdp  co2_per_unit_energy  coal_co2  \
150           0.052        0.093                0.177     0.004   

     coal_co2_per_capita  consumption_co2  consumption_co2_per_capita  \
150                  0.0              NaN                         NaN   

     consumption_co2_per_gdp  cumulative_cement_co2  cumulative_co2  \
150                      NaN                  1.969          72.449   

     cumulative_co2_including_luc  cumulative_coal_co2  \
150                       353.597               11.851   

     cumulative_flaring_co2  cumulative_gas_co2  cumulative_luc_co2  \
150                   5.934              13.897             794.416   

     cumulative_oil_co2  cumulative_other_co2  energy_per_capita  \
150              38.798                   NaN            302.595   

     energy_per_gdp  flaring_co2  flaring_co2_per_capita  gas_co2  \
150           0.524        0.022                   0.001    0.224   

     gas_co2_per_capita  ghg_excluding_lucf_per_capita  ghg_per_capita  \
150               0.011                          0.225           0.842   

     land_use_change_co2  land_use_change_co2_per_capita  methane  \
150                0.057                           0.003   11.643   

     methane_per_capita  nitrous_oxide  nitrous_oxide_per_capita  oil_co2  \
150               0.578          3.221                      0.16    0.788   

     oil_co2_per_capita  other_co2_per_capita  other_industry_co2  \
150               0.039                   NaN                 NaN   

     primary_energy_consumption  share_global_cement_co2  share_global_co2  \
150                       5.914                    0.001             0.004   

     share_global_co2_including_luc  share_global_coal_co2  \
150                           0.004                    0.0   

     share_global_cumulative_cement_co2  share_global_cumulative_co2  \
150                               0.011                        0.007   

     share_global_cumulative_co2_including_luc  \
150                                       0.02   

     share_global_cumulative_coal_co2  share_global_cumulative_flaring_co2  \
150                             0.002                                0.053   

     share_global_cumulative_gas_co2  share_global_cumulative_luc_co2  \
150                            0.011                            0.112   

     share_global_cumulative_oil_co2  share_global_cumulative_other_co2  \
150                            0.011                                NaN   

     share_global_flaring_co2  share_global_gas_co2  share_global_luc_co2  \
150                     0.006                 0.005                 0.001   

     share_global_oil_co2  share_global_other_co2  \
150                 0.008                     NaN   

     share_of_temperature_change_from_ghg  temperature_change_from_ch4  \
150                                 0.081                          0.0   

     temperature_change_from_co2  temperature_change_from_ghg  \
150                          0.0                        0.001   

     temperature_change_from_n2o  total_ghg  total_ghg_excluding_lucf  \
150                          0.0     16.955                     4.521   

     trade_co2  trade_co2_share Continent  
150        NaN              NaN      Asia

## 📊 Understanding the Descriptive Statistics Output
- **count**  
  Number of non-missing (non-null) values in that column.  
- **mean**  
  The arithmetic average of all values in the column.  
- **std (standard deviation)**  
  A measure of how much values vary around the mean  → High `std` means the values are very spread out, low `std` means they are clustered close to the mean.
- **min**  
  The smallest value observed in the column → Useful for spotting the lower boundary (e.g., countries with very small populations or GDPs).
- **25% (first quartile / Q1)**  
  The value below which 25% of the data falls → Helps understand the lower part of the distribution.
- **50% (median / Q2)**  
  The middle value when the data is sorted → Less sensitive to outliers than the mean.
- **75% (third quartile / Q3)**  
  The value below which 75% of the data falls → Helps understand the upper part of the distribution.
- **max**  
  The largest value observed in the column → Useful for identifying outliers or extreme cases (e.g., China or USA in emissions or GDP).


In [20]:
df_filtrado.groupby("Name")["co2"].agg(["mean","max","min"]).sort_values("min", ascending=False)


mean        max       min
Name                                              
United States     5562.299542   6132.183  4714.628
China             8352.364000  11902.503  3649.227
Russia            1633.733750   1815.925  1479.142
Japan             1207.055417   1315.298   988.785
India             1871.824917   3062.324   986.445
...                       ...        ...       ...
Christmas Island     0.000000      0.000     0.000
Kosovo               5.483375      8.816     0.000
Monaco                    NaN        NaN       NaN
San Marino                NaN        NaN       NaN
Vatican                   NaN        NaN       NaN

[219 rows x 3 columns]

In [21]:
df_filtrado.describe().T

count          mean           std           min  \
year                         5256.0  2.011500e+03  6.922845e+00  2.000000e+03   
population                   5208.0  3.292898e+07  1.306555e+08  5.010000e+02   
gdp                          3772.0  5.717035e+11  1.926304e+12  3.128536e+08   
cement_co2                   4982.0  6.148033e+00  4.495293e+01  0.000000e+00   
cement_co2_per_capita        4958.0  1.110131e-01  1.579547e-01  0.000000e+00   
...                             ...           ...           ...           ...   
temperature_change_from_n2o  4824.0  2.781924e-04  1.052102e-03  0.000000e+00   
total_ghg                    4704.0  2.493672e+02  9.485554e+02  5.000000e-03   
total_ghg_excluding_lucf     4680.0  1.972346e+02  8.754365e+02  4.000000e-03   
trade_co2                    2750.0  4.823127e-02  1.173473e+02 -1.532080e+03   
trade_co2_share              2750.0  2.615918e+01  5.904764e+01 -9.884900e+01   

                                      25%           50%           75%  \
year                         2.005750e+03  2.011500e+03  2.017250e+03   
population                   7.263345e+05  5.895870e+06  2.158235e+07   
gdp                          2.351140e+10  7.560725e+10  3.539220e+11   
cement_co2                   0.000000e+00  3.990000e-01  1.912250e+00   
cement_co2_per_capita        0.000000e+00  5.700000e-02  1.580000e-01   
...                                   ...           ...           ...   
temperature_change_from_n2o  0.000000e+00  0.000000e+00  0.000000e+00   
total_ghg                    9.280000e+00  4.707100e+01  1.297423e+02   
total_ghg_excluding_lucf     4.202750e+00  2.007250e+01  8.661400e+01   
trade_co2                   -1.320750e+00  2.239500e+00  9.583750e+00   
trade_co2_share             -5.143000e+00  1.192550e+01  3.826275e+01   

                                      max  
year                         2.023000e+03  
population                   1.438070e+09  
gdp                          2.696600e+13  
cement_co2                   8.499760e+02  
cement_co2_per_capita        1.243000e+00  
...                                   ...  
temperature_change_from_n2o  1.100000e-02  
total_ghg                    1.396892e+04  
total_ghg_excluding_lucf     1.357942e+04  
trade_co2                    6.541420e+02  
trade_co2_share              5.686350e+02  

[77 rows x 8 columns]

In [22]:
# Pandas shows very large or very small numbers in scientific notation (e.g., 5.717035e+11). 
# To convert this into the decimal format, take the coefficient (5.717035) and multiply it by ten to the power of eleven, which gives 571,703,500,000 (about 570 billion).

desc = df_filtrado.describe().T
desc_format = desc.applymap(lambda x: f"{x:,.2f}")
desc_format

/var/folders/fw/63lcqvrj3gj5_78zkmp2ql8r0000gp/T/ipykernel_63810/1234916815.py:5: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  desc_format = desc.applymap(lambda x: f"{x:,.2f}")


count                mean  \
year                         5,256.00            2,011.50   
population                   5,208.00       32,928,978.93   
gdp                          3,772.00  571,703,496,695.95   
cement_co2                   4,982.00                6.15   
cement_co2_per_capita        4,958.00                0.11   
...                               ...                 ...   
temperature_change_from_n2o  4,824.00                0.00   
total_ghg                    4,704.00              249.37   
total_ghg_excluding_lucf     4,680.00              197.23   
trade_co2                    2,750.00                0.05   
trade_co2_share              2,750.00               26.16   

                                              std             min  \
year                                         6.92        2,000.00   
population                         130,655,499.97          501.00   
gdp                          1,926,303,998,369.27  312,853,568.00   
cement_co2                                  44.95            0.00   
cement_co2_per_capita                        0.16            0.00   
...                                           ...             ...   
temperature_change_from_n2o                  0.00            0.00   
total_ghg                                  948.56            0.01   
total_ghg_excluding_lucf                   875.44            0.00   
trade_co2                                  117.35       -1,532.08   
trade_co2_share                             59.05          -98.85   

                                           25%                50%  \
year                                  2,005.75           2,011.50   
population                          726,334.50       5,895,869.50   
gdp                          23,511,404,032.00  75,607,248,896.00   
cement_co2                                0.00               0.40   
cement_co2_per_capita                     0.00               0.06   
...                                        ...                ...   
temperature_change_from_n2o               0.00               0.00   
total_ghg                                 9.28              47.07   
total_ghg_excluding_lucf                  4.20              20.07   
trade_co2                                -1.32               2.24   
trade_co2_share                          -5.14              11.93   

                                            75%                    max  
year                                   2,017.25               2,023.00  
population                        21,582,353.00       1,438,069,597.00  
gdp                          353,922,000,000.00  26,966,000,000,000.00  
cement_co2                                 1.91                 849.98  
cement_co2_per_capita                      0.16                   1.24  
...                                         ...                    ...  
temperature_change_from_n2o                0.00                   0.01  
total_ghg                                129.74              13,968.92  
total_ghg_excluding_lucf                  86.61              13,579.42  
trade_co2                                  9.58                 654.14  
trade_co2_share                           38.26                 568.63  

[77 rows x 8 columns]

## 📑 Explanation of Key Variables in the Dataset

- **gdp**  
  Gross Domestic Product (GDP), usually measured in constant international dollars.

- **cement_co2**  
  CO₂ emissions from cement production (absolute values).

- **cement_co2_per_capita**  
  CO₂ emissions from cement production per person (tons per capita).

- **temperature_change_from_n2o**  
  Estimated temperature change contribution from nitrous oxide (N₂O), measured in °C.

- **total_ghg**  
  Total greenhouse gas emissions, measured in CO₂-equivalent (includes CO₂, CH₄, N₂O, etc.).

- **total_ghg_excluding_lucf**  
  Total greenhouse gas emissions excluding land use change and forestry (LUCF).

- **trade_co2**  
  CO₂ emissions embedded in traded goods and services (can be negative if a country is a net importer of embodied emissions).

- **trade_co2_share**  
  Share (%) of trade-related CO₂ emissions relative to total CO₂ emissions.

## 📌 Key Patterns Identified

- **Strong global inequality in emissions**  
  → From `total_ghg`: the **mean = 249.37** is far above the **median = 47.07**, and the **max = 13,968.92**, indicating a few extreme emitters (e.g., China, US, India) dominate totals while most countries are much smaller.

- **Economic size and population as drivers of emissions**  
  → `population`: **mean = 32,928,978.93** vs **median = 5,895,869.50**.  
  → `gdp`: **mean = 571,703,496,695.95** vs **median = 75,607,248,896.00**.  
  These gaps show that a small number of very large economies/populations skew global averages, linking economic and demographic size to emission levels.

- **Median GHG emissions much lower than the mean**  
  → Again in `total_ghg`: **median = 47.07** vs **mean = 249.37**, confirming a skewed distribution where most countries emit relatively little while a few contribute disproportionately.

- **Fossil fuels dominate, cement is secondary**  
  → `cement_co2` has **max = 849.98** (and `cement_co2_per_capita` **max = 1.24**), which is modest compared to fossil-fuel emissions columns, underscoring the dominance of coal, oil, and gas while cement remains smaller but relevant.

- **Trade shifts responsibility for emissions**  
  → `trade_co2`: **min = −1,532.08**, **max = 654.14**; `trade_co2_share`: **min = −98.85**, **max = 568.63**.  
  Negative values indicate **net importers** of embodied CO₂, positive values indicate **net exporters**, revealing asymmetry between production- and consumption-based emissions.


In [23]:
df_filtrado.to_csv('data_clean_country.csv')

In [24]:
regions = [
    'Africa', 'Africa (GCP)',
    'Asia', 'Asia (GCP)', 'Asia (excl. China and India)',
    'Central America (GCP)', 'Europe', 'Europe (GCP)',
    'Europe (excl. EU-27)', 'Europe (excl. EU-28)',
    'Middle East (GCP)', 'North America', 'North America (GCP)',
    'North America (excl. USA)', 'Oceania', 'Oceania (GCP)',
    'Ryukyu Islands', 'Ryukyu Islands (GCP)', 'South America',
    'South America (GCP)', 'World', 'European Union (27)',
    'European Union (28)', 'High-income countries',
    'Least developed countries (Jones et al.)', 'Low-income countries',
    'Lower-middle-income countries', 'Non-OECD (GCP)', 'OECD (GCP)',
    'OECD (Jones et al.)', 'Upper-middle-income countries'
]

df_regions = filter_year[filter_year["Name"].isin(regions)]


In [25]:
df_regions.to_csv('data_clean_regions.csv')